hnswalg.cpp

```diff
@@ -233,101 +251,181 @@ dist_t fstdistfunc_scalar(const coord_t *x, const coord_t *y, size_t n)
 
 }
 
+dist_t dist_manhattan_scalar(const coord_t *x, const coord_t *y, size_t n)
+{
+    dist_t 	distance = 0.0;
+
+    #pragma clang loop vectorize(enable)
+    for (size_t i = 0; i < n; i++)
+    {
+        dist_t diff = std::abs(x[i] - y[i]);
+        distance += diff;
+    }
+    return distance;
+}
+
 #ifdef __x86_64__
 #include <immintrin.h>
 
 __attribute__((target("avx2")))
-dist_t fstdistfunc_avx2(const coord_t *x, const coord_t *y, size_t n)
+dist_t dist_manhattan_avx2(const coord_t *x, const coord_t *y, size_t n)
 {
-    const size_t TmpResSz = sizeof(__m256) / sizeof(float);
-    float PORTABLE_ALIGN32 TmpRes[TmpResSz];
-    size_t qty16 = n / 16;
-    const float *pEnd1 = x + (qty16 * 16);
-    __m256 diff, v1, v2;
-    __m256 sum = _mm256_set1_ps(0);
-
-    while (x < pEnd1) {
-        v1 = _mm256_loadu_ps(x);
-        x += 8;
-        v2 = _mm256_loadu_ps(y);
-        y += 8;
-        diff = _mm256_sub_ps(v1, v2);
-        sum = _mm256_add_ps(sum, _mm256_mul_ps(diff, diff));
-
-        v1 = _mm256_loadu_ps(x);
-        x += 8;
-        v2 = _mm256_loadu_ps(y);
-        y += 8;
-        diff = _mm256_sub_ps(v1, v2);
-        sum = _mm256_add_ps(sum, _mm256_mul_ps(diff, diff));
+    const size_t elts_per_vector = sizeof(__m256) / sizeof(coord_t);
+    float partial_result[elts_per_vector];
+    __m256 accumulator = _mm256_setzero_ps();
+    __m256 vecMask = _mm256_set1_ps(-0.0f);
+    for(size_t i = 0; i < n; i += elts_per_vector)
+    {
+        __m256 vecX = _mm256_loadu_ps(x + i);
+        __m256 vecY = _mm256_loadu_ps(y + i);
+        __m256 vecSub = _mm256_sub_ps(vecX, vecY);
+        __m256 vecAbs = _mm256_andnot_ps(vecMask, vecSub);
+        accumulator = _mm256_add_ps(accumulator, vecAbs);
+    }
+    _mm256_storeu_ps(partial_result, accumulator);
+    float res1 = partial_result[0] + partial_result[4];
+    float res2 = partial_result[1] + partial_result[5];
+    float res3 = partial_result[2] + partial_result[6];
+    float res4 = partial_result[3] + partial_result[7];
+    res1 += res3;
+    res2 += res4;
+    res1 += res2;
+
+    size_t tail_size = n % elts_per_vector;
+    size_t tail = n - tail_size;
+    for(int i = 0; i < tail_size; i++)
+    {
+        dist_t dist = x[tail + i] - y[tail + i];
```

Comment:
Ok, I'm not familiar with avx2, does this convert automatically to the absolute value of the difference?

Reply:
Ah you're right :) 

When you do AVX2 you have to have two loops, one for the most of the values (e.g. we process 16 values per loop, so the first multiple-of-16 values) and then one for the tail (the last few elements). I updated the main loop but not the tail.